contrib_bots/lib/converter.py

```diff
@@ -0,0 +1,268 @@
+    It looks for messages of the format
+    '@convert <number> <original-unit> <new-unit>'
+    The message '@convert help' posts a short description of how to use
+    the plugin, along with a list of all supported units.
+    '''
+
+    def usage(self):
+        return '''
+               This plugin allows users to make conversions between
+               various units, e.g. Celsius to Fahrenheit,
+               or kilobytes to gigabytes. It looks for messages of
+               the format '@convert <number> <original-unit> <new-unit>'
+               The message '@convert help' posts a short description of
+               how to use the plugin, along with a list of
+               all supported units.
+               '''
+
+    def triage_message(self, message, client):
+        return ('@convert' in message['content'])
+
+    def handle_message(self, message, client, state_handler):
+        content = message['content']
+
+        words = content.lower().split()
+        convert_indexes = [i for i, word in enumerate(words) if word == "@convert"]
+        results = []
+
+        for convert_index in convert_indexes:
+            if (convert_index + 1) < len(words) and words[convert_index + 1] == 'help':
+                results.append(HELP_MESSAGE)
+                continue
+            if (convert_index + 3) < len(words):
+                number = words[convert_index + 1]
+                unit_from = ALIASES.get(words[convert_index + 2], words[convert_index + 2])
+                unit_to = ALIASES.get(words[convert_index + 3], words[convert_index + 3])
+                exponent = 0
+
+                if not isfloat(number):
+                    results.append(number + ' is not a valid number. ' +
+                                   'Enter `@convert help` for help on using the converter.')
+                    continue
+
+                number = float(number)
+                number_res = copy.copy(number)
+
+                for key, exp in PREFIXES.items():
+                    if unit_from.startswith(key):
+                        exponent += exp
+                        unit_from = unit_from[len(key):]
+                    if unit_to.startswith(key):
+                        exponent -= exp
+                        unit_to = unit_to[len(key):]
+
+                uf_to_std = UNITS.get(unit_from, False)
+                ut_to_std = UNITS.get(unit_to, False)
+
+                if uf_to_std == False:
+                    results.append(unit_from + ' is not a valid unit.')
+                if ut_to_std == False:
+                    results.append(unit_to + ' is not a valid unit.')
+                if uf_to_std == False or ut_to_std == False:
+                    continue
+
+                base_unit = uf_to_std[2]
+                if uf_to_std[2] != ut_to_std[2]:
+                    results.append(unit_to + ' and ' + unit_from +
+                                   ' are not from the same category.')
```

Comment:
Also, you might want to add `Enter @convert help for help on using the converter` here.